### PR TITLE
CLI should take in a runtime config

### LINKED ISSUE #2919: CLI needs to accept a runtime config file and path
The CLI needs an `AtlasDbConfig` object that is constructed through a file and a path. Since we moved the `timelock` config out of the `AtlasDbConfig` config to the `AtlasDbRuntimeConfig`, the CLI doesn't know anymore how to talk to timelock, failing when trying to run any command. We need to make the CLI also be able to build an `AtlasDbRuntimeConfig` object, from a file and a path. I.e. add a `--runtime-config` and a `--runtime-config-root-path` parameters.

### PR DESCRIPTION
**Pull request by hsaraogi**
**Goals (and why)**: CLI needs to take a runtime config parameter. Fixes #2919.

**Implementation Description (bullets)**: CLI can take a run time config and start up a service with it.

**Concerns (what feedback would you like?)**: This is breaking as we hard-fail if the runtime config isnt specified. The alternative is to fall back to the default runtime config, but I dont think it is safe, especially with timelock as you could have a cluster specified in the runtime config.

**Where should we start reviewing?**: `AbstractCommand`

**Priority (whenever / two weeks / yesterday)**: soon.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2937)

### CODE REVIEW

**Comment by hsaraogi** [comment]
@fsamuel-bs I think the config root is still `atlasdb`/`atlas`.

File: docs/source/cluster_management/clis.rst [hunk 1]
```diff
@@ -18,6 +18,52 @@ The CLI now comes prebuilt as an SLS distribution to be used with any AtlasDB-ba
...
+
+    *    - ``--config-root``
+         - The AtlasDB config root for your install config. The CLI will by default look for ``atlasdb``/``atlas`` blocks in the config, but services can name these blocks arbitrarily.
+           This can usually be found by cracking open the install config file and searching for parameters defined in
+           `Atlas Install config <https://github.com/palantir/atlasdb/blob/develop/atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java>`__.
+           The parent of these parameters will be the install config root.
+         - No
+
+    *    - ``--runtime-config-root``
+         - The AtlasDB config root for your runtimr config. The CLI will by default look for ``atlasdb``/``atlas`` blocks in the config, but services can name these blocks arbitrarily.
```
Comment (fsamuel-bs) [inline docs/source/cluster_management/clis.rst]: runtimr -> runtime.

File: docs/source/cluster_management/clis.rst [hunk 2]
```diff
@@ -18,6 +18,52 @@ The CLI now comes prebuilt as an SLS distribution to be used with any AtlasDB-ba
...
+           This can usually be found by cracking open the install config file and searching for parameters defined in
```
Comment (fsamuel-bs) [inline docs/source/cluster_management/clis.rst]: install -> runtime

File: docs/source/cluster_management/clis.rst [hunk 3]
```diff
@@ -18,6 +18,52 @@ The CLI now comes prebuilt as an SLS distribution to be used with any AtlasDB-ba
...
+           `Atlas Runtime config <https://github.com/palantir/atlasdb/blob/develop/atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbRuntimeConfig.java>`__.
+           The parent of these parameters will be the install config root. If you do not have any AtlasDB runtime config,
```
Comment (fsamuel-bs) [inline docs/source/cluster_management/clis.rst]: install -> runtime

**Comment by hsaraogi** [comment]
Smoke tested, this now connects to timelock on providing a runtime config. Found a couple of fatal bugs and fixed them.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 4]
```diff
@@ -453,22 +451,22 @@ private static PersistentLockService createAndRegisterPersistentLockService(
...
-    public static LockAndTimestampServices createLockAndTimestampServices(
+    public static LockAndTimestampServices createLockAndTimestampServicesForCli(
             AtlasDbConfig config,
+            java.util.function.Supplier<AtlasDbRuntimeConfig> runtimeConfigSupplier,
             Consumer<Object> env,
             com.google.common.base.Supplier<LockService> lock,
-            com.google.common.base.Supplier<TimestampService> time) {
+            com.google.common.base.Supplier<TimestampService> time,
+            TimestampStoreInvalidator invalidator,
+            String userAgent) {
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: It's a shame we didn't upgrade Guava to 21! (Guava Supplier becomes a Java Supplier in 21)

File: atlasdb-dropwizard-bundle/src/main/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtils.java [hunk 5]
```diff
@@ -43,18 +46,49 @@ private AtlasDbCommandUtils() {
...
-                "Your server configuration file must have a leader or timelock block. For instructions on how to do "
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static AtlasDbConfig convertServerConfigToClientConfig(AtlasDbConfig installConfig,
+            Optional<AtlasDbRuntimeConfig> runtimeConfig) {
+        Preconditions.checkArgument(installConfig.leader().isPresent()
+                        || installConfig.timelock().isPresent()
+                        || runtimeConfigHasTimeLockBlock(runtimeConfig),
+                "Your server configuration file must have a leader — in the install time config — "
+                        + "or timelock block — in either the install or the runtime config. "
+                        + "For instructions on how to do "
```
Comment (jeremyk-91) [inline atlasdb-dropwizard-bundle/src/main/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtils.java]: Note to self: Inconsistencies here are handled by TMs itself.

File: atlasdb-dropwizard-bundle/src/main/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtils.java [hunk 6]
```diff
@@ -43,18 +46,49 @@ private AtlasDbCommandUtils() {
...
+        if (runtimeConfigHasTimeLockBlock(runtimeConfig)) {
+            //noinspection ConstantConditions
+            return convertRuntimeConfigWithTimeLock(installConfig, runtimeConfig.get());
+        }
+
+        return convertConfigWithLeaderBlockToClientConfig(installConfig);
+    }
+
+    private static boolean runtimeConfigHasTimeLockBlock(Optional<AtlasDbRuntimeConfig> runtimeConfig) {
+        return runtimeConfig.isPresent() && runtimeConfig.get().timelockRuntime().isPresent();
```
Comment (jeremyk-91) [inline atlasdb-dropwizard-bundle/src/main/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtils.java]: or `runtimeConfig.map(AtlasDbRuntimeConfig::timelockRuntime).isPresent()` 😄

File: atlasdb-dropwizard-bundle/src/test/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtilsTest.java [hunk 7]
```diff
@@ -78,46 +83,64 @@ public void setUp() {
...
     public void clientConfigMatchesServerConfigForTimelock() {
-        AtlasDbConfig clientConfig = AtlasDbCommandUtils.convertServerConfigToClientConfig(timeLockConfig);
+        AtlasDbConfig clientConfig = AtlasDbCommandUtils.convertServerConfigToClientConfig(timeLockConfig,
+                Optional.of(EMPTY_RUNTIME_CONFIG));
+
+        assertThat(clientConfig).isEqualTo(timeLockConfig);
+    }
+
+    @Test
+    public void clientConfigMatchesRuntimeConfigForTimelock() {
```
Comment (jeremyk-91) [inline atlasdb-dropwizard-bundle/src/test/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtilsTest.java]: 👍

File: docs/source/cluster_management/clis.rst [hunk 8]
```diff
@@ -18,6 +18,53 @@ The CLI now comes prebuilt as an SLS distribution to be used with any AtlasDB-ba
 You can find the versions `here <https://palantir.bintray.com/releases/com/palantir/atlasdb/atlasdb-cli-distribution/>`__ for download.
 You can use the CLI by unpacking the tar ball and running the executable in ``service/bin/atlasdb-cli``.
 
+CLI Options
+===========
```
Comment (jeremyk-91) [inline docs/source/cluster_management/clis.rst]: Thanks a lot for doing this!

nit: maybe worth clarifying as 'General CLI options' or something like that?

File: docs/source/cluster_management/clis.rst [hunk 9]
```diff
@@ -18,6 +18,53 @@ The CLI now comes prebuilt as an SLS distribution to be used with any AtlasDB-ba
...
+         - Use it if you want to provide the AtlasDB config as a string/json.
+           The config string should immediately follow the option.
+         - No
+
+    *    - ``--config-root``
+         - The AtlasDB config root for your install config. The CLI will by default look for ``/atlasdb``/``/atlas`` blocks in the config, but services can name these blocks arbitrarily.
+           This can usually be found by cracking open the install config file and searching for parameters defined in
+           `Atlas Install config <https://github.com/palantir/atlasdb/blob/develop/atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java>`__.
+           The parent of these parameters will be the install config root.
+           Note that you have to provide the root with a '/' as the first character.
```
Comment (jeremyk-91) [inline docs/source/cluster_management/clis.rst]: ❤️  this has caused a lot of... surprises.

File: atlasdb-dropwizard-bundle/src/main/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtils.java [hunk 10]
```diff
@@ -43,18 +46,49 @@ private AtlasDbCommandUtils() {
...
+        if (runtimeConfigHasTimeLockBlock(runtimeConfig)) {
+            //noinspection ConstantConditions
+            return convertRuntimeConfigWithTimeLock(installConfig, runtimeConfig.get());
+        }
+
+        return convertConfigWithLeaderBlockToClientConfig(installConfig);
+    }
+
+    private static boolean runtimeConfigHasTimeLockBlock(Optional<AtlasDbRuntimeConfig> runtimeConfig) {
+        return runtimeConfig.flatMap(AtlasDbRuntimeConfig::timelockRuntime).isPresent();
```
Comment (fsamuel-bs) [inline atlasdb-dropwizard-bundle/src/main/java/com/palantir/atlasdb/dropwizard/commands/AtlasDbCommandUtils.java]: @jeremyk-91 actually, it needs to be a `flatMap`.

**Pull request merged**